messagequeue/messagequeue.go

```diff
@@ -16,63 +21,172 @@ var log = logging.Logger("bitswap")
-
-	// do not touch out of run loop
-	wl                    *wantlist.SessionTrackedWantlist
-	nextMessage           bsmsg.BitSwapMessage
-	nextMessageLk         sync.RWMutex
+	ctx              context.Context
+	p                peer.ID
+	network          MessageNetwork
+	maxMessageSize   int
+	sendErrorBackoff time.Duration
+
+	signalWorkReady func()
+	outgoingWork    chan struct{}
+	done            chan struct{}
+
+	// Take lock whenever any of these variables are modified
+	wllock    sync.Mutex
+	bcstWants *recallWantlist
+	peerWants *recallWantlist
+	cancels   *cid.Set
+	priority  int
+
+	// Dont touch any of these variables outside of run loop
 	sender                bsnet.MessageSender
 	rebroadcastIntervalLk sync.RWMutex
 	rebroadcastInterval   time.Duration
 	rebroadcastTimer      *time.Timer
 }
 
+// recallWantlist keeps a list of current wants, and a list of all wants that
+// have ever been requested
+type recallWantlist struct {
+	allWants *bswl.Wantlist
+	current  *bswl.Wantlist
+}
+
+func newRecallWantList() *recallWantlist {
+	return &recallWantlist{
+		allWants: bswl.New(),
+		current:  bswl.New(),
+	}
+}
+
+// Add want to both the current list and the list of all wants
+func (r *recallWantlist) Add(c cid.Cid, priority int, wtype pb.Message_Wantlist_WantType) {
+	r.allWants.Add(c, priority, wtype)
+	r.current.Add(c, priority, wtype)
+}
+
+// Remove wants from both the current list and the list of all wants
+func (r *recallWantlist) Remove(c cid.Cid) {
+	r.allWants.Remove(c)
+	r.current.Remove(c)
+}
+
+// Remove wants by type from both the current list and the list of all wants
+func (r *recallWantlist) RemoveType(c cid.Cid, wtype pb.Message_Wantlist_WantType) {
+	r.allWants.RemoveType(c, wtype)
+	r.current.RemoveType(c, wtype)
+}
+
 // New creats a new MessageQueue.
 func New(ctx context.Context, p peer.ID, network MessageNetwork) *MessageQueue {
-	return &MessageQueue{
+	return newMessageQueue(ctx, p, network, maxMessageSize, sendErrorBackoff)
+}
+
+// This constructor is used by the tests
+func newMessageQueue(ctx context.Context, p peer.ID, network MessageNetwork, maxMsgSize int, sendErrorBackoff time.Duration) *MessageQueue {
+	mq := &MessageQueue{
 		ctx:                 ctx,
-		wl:                  wantlist.NewSessionTrackedWantlist(),
-		network:             network,
 		p:                   p,
+		network:             network,
+		maxMessageSize:      maxMsgSize,
+		bcstWants:           newRecallWantList(),
+		peerWants:           newRecallWantList(),
+		cancels:             cid.NewSet(),
 		outgoingWork:        make(chan struct{}, 1),
 		done:                make(chan struct{}),
 		rebroadcastInterval: defaultRebroadcastInterval,
+		sendErrorBackoff:    sendErrorBackoff,
+		priority:            maxPriority,
 	}
+
+	// Apply debounce to the work ready signal (which triggers sending a message)
+	debounced := debounce.New(sendMessageDebounce)
+	mq.signalWorkReady = func() { debounced(mq.onWorkReady) }
```

Comment:
hmm yeah that's a good point - maybe it would make sense to abstract this into a debounce function that takes two parameters:
- debounce time
  wait at least this interval after the last invocation before calling the target function
- max time
  if we're still debouncing after this amount of time, don't wait any more, call the target function anyway

Reply:
That would work. Ideally, we'd look at the message _size_ (as well?) but time is probably good enough for now.

(I'm also slightly concerned about all the timer allocations but that's something we can fix if it shows up in profiles)

Reply:
Added a TODO checkbox at the top of this PR